sentry-core/src/main/java/io/sentry/core/transport/HttpTransport.java

```diff
@@ -297,46 +277,83 @@ public boolean isRetryAfter(final @NotNull String itemType) {
     int responseCode = -1;
 
     try (final OutputStream outputStream = connection.getOutputStream();
-        final Writer writer = new BufferedWriter(new OutputStreamWriter(outputStream, UTF_8))) {
+        final GZIPOutputStream gzip = new GZIPOutputStream(outputStream);
+        final Writer writer = new BufferedWriter(new OutputStreamWriter(gzip, UTF_8))) {
       serializer.serialize(envelope, writer);
 
-      // need to also close the input stream of the connection
-      connection.getInputStream().close();
       options.getLogger().log(DEBUG, "Envelope sent successfully.");
       return TransportResult.success();
     } catch (IOException e) {
-      try {
-        responseCode = connection.getResponseCode();
+      final TransportResult errorResult = getResultFromErrorCode(connection, e);
+      responseCode = errorResult.getResponseCode();
+      return errorResult;
+    } catch (Exception e) {
+      return getDefaultErrorResult(e);
+    } finally {
+      updateRetryAfterLimits(connection, responseCode);
 
-        // TODO: 403 part of the protocol?
-        if (responseCode == HttpURLConnection.HTTP_FORBIDDEN) {
-          options
-              .getLogger()
-              .log(DEBUG, "Envelope was rejected by the Sentry server due to a filter.");
-        }
-        logErrorInPayload(connection);
-        return TransportResult.error(responseCode);
-      } catch (IOException responseCodeException) {
-        // this should not stop us from continuing. We'll just use -1 as response code.
-        options
-            .getLogger()
-            .log(WARNING, "Failed to obtain response code while analyzing event send failure.", e);
-      }
+      closeAndDisconnectConnection(connection);
+    }
+  }
+
+  /**
+   * Closes the Response stream and disconnect the connection
+   *
+   * @param connection the HttpURLConnection
+   */
+  private void closeAndDisconnectConnection(final @NotNull HttpURLConnection connection) {
+    try {
+      connection.getInputStream().close();
+    } catch (IOException e) {
+      options.getLogger().log(ERROR, e, "Error while closing the connection.");
+    }
+    connection.disconnect();
```

Comment:
`m`: I think currently it doesn't make a big difference, because we already call this in `finally` and we don't have any return statements in this function, but wouldn't it be safer to wrap this in `finally`?

Reply:
`h`: I think it is not best practice to disconnect the connection for every single envelope and event that we send. It is better to keep the connection alive for some time. For each new connection, a handshake needs to be made, a TCP connection needs to be set up, etc. Read more about this [here](https://docs.oracle.com/javase/6/docs/technotes/guides/net/http-keepalive.html). Maybe we can keep the connection open for some time, but I guess this is out of scope in this PR. I just wanted to mention it. 
Maybe we also have good reasons not to do so that I am not aware of.

Reply:
1st comment: this method is called in a `finally` block, but we have 2 calls, send(envelope) and send(event), so extracted to a method, DRY.

Reply:
2nd comment, I generally agree, but indeed is out of this scope, worth mentioning that keeping a connection open for some time also has downsides.
this will be worked on when sentry-java lands to this repo.

Reply:
> 1st comment: this method is called in a `finally` block, but we have 2 calls, send(envelope) and send(event), so extracted to a method, DRY.

Yep, but how do we know that everybody is calling this in a `finally` block in the future? Adding it makes it harder to misuse the method. That's [poka yoke](https://en.wikipedia.org/wiki/Poka-yoke).

Reply:
@philipphofmann it's a private method that closes the response stream and disconnect the connection, calling it on `finally` is just "a good approach", if you prefer to call it in all the catch blocks, it'll behave exactly the same, but then you repeat even more code.
So I don't get your point here, do you prefer to duplicate this code in both finally blocks?

Reply:
That would be my suggestion.
```suggestion
    finally {
      connection.disconnect();
    }
```
As already said it doesn't make a difference right now. We can keep it like it this. `YAGNI`

Reply:
@philipphofmann ok got it, I thought you were mentioning the other 2 finally blocks of send(event/envelope)

Reply:
yep it does make sense, will do it, thx